helios-client/src/main/java/com/spotify/helios/client/DefaultRequestDispatcher.java

```diff
@@ -0,0 +1,298 @@
+                                    final Map<String, List<String>> headers)
+      throws URISyntaxException, IOException, TimeoutException, InterruptedException,
+             HeliosException {
+    final long deadline = currentTimeMillis() + RETRY_TIMEOUT_MILLIS;
+    final int offset = ThreadLocalRandom.current().nextInt();
+    while (currentTimeMillis() < deadline) {
+      final List<URI> endpoints = endpointSupplier.get();
+      if (endpoints.isEmpty()) {
+        throw new RuntimeException("failed to resolve master");
+      }
+      log.debug("endpoint uris are {}", endpoints);
+
+      // Resolve hostname into IPs so client will round-robin and retry for multiple A records.
+      // Keep a mapping of IPs to hostnames for TLS verification.
+      final List<URI> ipEndpoints = Lists.newArrayList();
+      final Map<URI, URI> ipToHostnameUris = Maps.newHashMap();
+
+      for (final URI hnUri : endpoints) {
+        try {
+          final InetAddress[] ips = InetAddress.getAllByName(hnUri.getHost());
+          for (final InetAddress ip : ips) {
+            final URI ipUri = new URI(
+                hnUri.getScheme(), hnUri.getUserInfo(), ip.getHostAddress(), hnUri.getPort(),
+                hnUri.getPath(), hnUri.getQuery(), hnUri.getFragment());
+            ipEndpoints.add(ipUri);
+            ipToHostnameUris.put(ipUri, hnUri);
+          }
+        } catch (UnknownHostException e) {
+          log.warn("Unable to resolve hostname {} into IP address: {}", hnUri.getHost(), e);
+        }
+      }
+
+      for (int i = 0; i < ipEndpoints.size() && currentTimeMillis() < deadline; i++) {
+        final URI ipEndpoint = ipEndpoints.get(positive(offset + i) % ipEndpoints.size());
+        final String fullpath = ipEndpoint.getPath() + uri.getPath();
+
+        final String scheme = ipEndpoint.getScheme();
+        final String host = ipEndpoint.getHost();
+        final int port = ipEndpoint.getPort();
+        if (!VALID_PROTOCOLS.contains(scheme) || host == null || port == -1) {
+          throw new HeliosException(String.format(
+              "Master endpoints must be of the form \"%s://heliosmaster.domain.net:<port>\"",
+              VALID_PROTOCOLS_STR));
+        }
+
+        final URI realUri = new URI(scheme, host + ":" + port, fullpath, uri.getQuery(), null);
+        try {
+          log.debug("connecting to {}", realUri);
+          return connect0(realUri, method, entity, headers,
+                          ipToHostnameUris.get(ipEndpoint).getHost());
+        } catch (ConnectException | SocketTimeoutException | UnknownHostException e) {
+          // UnknownHostException happens if we can't resolve hostname into IP address.
+          // UnknownHostException's getMessage method returns just the hostname which is a useless
+          // message, so log the exception class name to provide more info.
+          log.debug(e.getClass().getSimpleName() + " - " + e.getMessage());
```

Comment:
almost :) e.toString will print the full class name, with the package. This has not been changed in this patch, just moved around but I can change it.